crypto/hpke/hpke.c

```diff
@@ -0,0 +1,1537 @@
+                   const unsigned char *aad, size_t aadlen,
+                   const unsigned char *pt, size_t ptlen)
+{
+    int erv = 1;
+    unsigned char seqbuf[12];
+    size_t seqlen = 0;
+    unsigned char nonce[12];
+
+    if (ctx == NULL || ct == NULL || ctlen == NULL || *ctlen == 0
+        || pt == NULL || ptlen == 0) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_PASSED_INVALID_ARGUMENT);
+        return 0;
+    }
+    if (ctx->key == NULL || ctx->nonce == NULL) {
+        /* need to have done an encap first, info can be NULL */
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+        return 0;
+    }
+
+    if (ctx->noncelen > sizeof(nonce)) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+        return 0;
+    }
+    seqlen = hpke_seq2buf(ctx->seq, seqbuf, sizeof(seqbuf));
+    if (seqlen == 0) {
+        erv = 0;
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        goto err;
+    } else {
+        size_t sind;
+        unsigned char cv;
+
+        if (seqlen > ctx->noncelen) {
+            erv = 0;
+            ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+        /* non constant time - does it matter? maybe no */
+        for (sind = 0; sind != ctx->noncelen; sind++) {
+            nonce[ctx->noncelen - 1 - sind] =
+                ctx->nonce[ctx->noncelen - 1 - sind];
+            if (sind < seqlen) {
+                cv = seqbuf[seqlen - 1 - (sind % seqlen)];
+            } else {
+                cv = 0x00;
+            }
+            nonce[ctx->noncelen - 1 - sind] ^= cv;
+        }
+    }
+
+    erv = hpke_aead_enc(ctx->libctx, ctx->propq, ctx->suite,
+                        ctx->key, ctx->keylen, nonce, ctx->noncelen,
+                        aad, aadlen, pt, ptlen, ct, ctlen);
+    if (erv != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    } else {
+        ctx->seq++;
+    }
+err:
+    OPENSSL_cleanse(nonce, ctx->noncelen);
+    return 1;
+}
+
+int OSSL_HPKE_open(OSSL_HPKE_CTX *ctx,
+                   unsigned char *pt, size_t *ptlen,
+                   const unsigned char *aad, size_t aadlen,
+                   const unsigned char *ct, size_t ctlen)
+{
+    int erv = 1;
+    unsigned char seqbuf[12];
+    size_t seqlen = 0;
+    unsigned char nonce[12];
+
+    if (ctx == NULL || pt == NULL || ptlen == NULL || *ptlen == 0
+        || ct == NULL || ctlen == 0) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_PASSED_INVALID_ARGUMENT);
+        return 0;
+    }
+    if (ctx->key == NULL || ctx->nonce == NULL) {
+        /* need to have done an encap first, info can be NULL */
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+        return 0;
+    }
+
+    if (ctx->noncelen > sizeof(nonce)) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+        return 0;
+    }
+    seqlen = hpke_seq2buf(ctx->seq, seqbuf, sizeof(seqbuf));
```

Comment:
See above.. This whole blob of code should be a single simple function that is called.

Reply:
handled with the above (will mark this resolved and see what CI does before doing so with the above)